### ISSUE TITLE
load() produces incoherent output for VLM model checkpoints used as text-only

### ISSUE THREAD

**Issue by Thump604** [ml-explore/mlx-lm#1197]
## Problem

`mlx_lm.load()` loads VLM model checkpoints (e.g. `Qwen3.6-35B-A3B-VLM-MTP-8bit`, `Qwen3.5-122B-A10B-VLM-MTP-5bit`) and returns a model object, but generation produces incoherent/garbage output. The model class resolves correctly (`qwen3_5_moe.Model`) and no error is raised, but the text is nonsensical.

These models are distributed as VLM checkpoints with vision components. When `load()` initializes the text model from these checkpoints, something in the weight mapping or initialization produces a broken text model.

## Reproduction

```python
from mlx_lm import load, generate

model, tokenizer = load("Qwen3.6-35B-A3B-VLM-MTP-8bit")  # local path
print(type(model).__name__)  # "Model" (qwen3_5_moe.Model) -- correct class

result = generate(model, tokenizer, "What is 2+2?", max_tokens=50)
print(result)
# Garbage: "5inaatroEB No... ignavierte... 酒店预订..."
```

Non-VLM checkpoints of the same architecture work correctly:

```python
model, tokenizer = load("mlx-community/Qwen3.5-2B-OptiQ-4bit")
result = generate(model, tokenizer, "What is 2+2?", max_tokens=50)
# Correct output
```

## Impact

Most Qwen 3.5/3.6 models distributed for Apple Silicon are VLM variants. Users who download these models and try to use them for text generation through `load()` + `generate()` get silent failures (garbage output, no error).

## Expected behavior

Either:
1. `load()` correctly initializes the text model from VLM checkpoints (skipping vision weights), or
2. `load()` raises a clear error when a VLM checkpoint is detected, directing users to mlx-vlm

## Environment

- mlx-lm: upstream main (ed1fca4)
- Models tested: Qwen3.6-35B-A3B-VLM-MTP-8bit, Qwen3.5-122B-A10B-VLM-MTP-5bit
- Working: Qwen3-1.7B-MLX-8bit, Qwen3.5-2B-OptiQ-4bit (both non-VLM)

**Comment by raullenchai** [comment]
We ran into this in rapid-mlx and did some extra isolation that might help narrow the root cause. Confirming @Thump604's report on `mlx-community/Qwen3.6-35B-A3B-*` (both the 4-bit and 8-bit MLX quants), with a few things we were able to rule out:

**Not the weights or the download.** The cached safetensors are byte-identical to the HF originals (sha256 match), and the *same* weights generate perfectly coherent text when loaded through `mlx-vlm` — so a correct forward for this checkpoint exists; the breakage is specific to the `mlx-lm` text-only path.

**Not a weight-mapping / missing-tensor issue.** `mlx_lm.load()` on `Qwen3.6-35B-A3B-4bit` fills every one of the 1757 text params (0 randomly-initialized slots); the only 333 dropped keys are `vision_tower.*`, correctly filtered. Generation is still garbage.

**Not config-driven.** Symlinking the weights and swapping in 4 config variants — including making the `text_config` byte-match a *working* sibling (`Qwen3.5-35B-A3B`) — produced identical garbage every time.

**Not (only) the mixed-bit `sanitize` gap from #1214.** The 8-bit quant garbles too, and its per-tensor `quantization` overrides are all 8-bit (== the global default), so remapping them is a no-op. #1214 may compound for the 4-bit mixed build, but it can't be the sole cause here.

**Rolling `mlx-lm` back doesn't help.** 0.31.2 → identical garbage; 0.30.7 → load fails (vision_tower filtering regressed to pre-#946 behavior).

The sibling that *works* under the exact same `mlx_lm.models.qwen3_5_moe.Model` class + config + loader is `Qwen3.5-35B-A3B`, so whatever diverges is data-dependent in the Qwen3.6 checkpoints and is handled correctly by `mlx-vlm`'s implementation of the same architecture. Diffing `mlx-lm`'s vs `mlx-vlm`'s `qwen3_5` forward is probably the fastest path to the fix. Happy to help test a patch.

**Comment by raullenchai** [comment]
Found the exact cause — it's in the **loader (`sanitize`), not the forward pass**: `mlx-lm` applies an erroneous `+1.0` RMSNorm-weight shift to this checkpoint when it shouldn't.

**Evidence.** Loading `mlx-community/Qwen3.6-35B-A3B-4bit` through `mlx_lm.load()` and through `mlx_vlm.load()` and diffing every text-model tensor: the *only* thing that differs is the norm weights, and they differ by **exactly +1.0**:

| tensor | mlx-lm mean | mlx-vlm mean | Δ |
|---|---|---|---|
| `*.input_layernorm.weight` (40/40) | +2.06 | +1.06 | +1.0078 |
| `*.post_attention_layernorm.weight` (40/40) | +2.15 | +1.15 | +1.0078 |
| `*.self_attn.q_norm.weight` (10/10) | +2.49 | +1.49 | +1.0078 |
| `*.self_attn.k_norm.weight` (10/10) | +2.54 | +1.54 | +1.0078 |
| `model.norm.weight` (1/1) | +3.63 | +2.63 | +1.0156 |

Every other tensor (embeddings, attn/MoE projections, `linear_attn` incl. `conv1d`, all quant scales/biases) is bit-identical between the two loaders.

**Why it misfires.** `qwen3_5.py`'s `sanitize` shifts norm weights by `+1.0` when `should_shift_norm_weights = has_mtp_weights or has_unsanitized_conv1d`. For this checkpoint neither holds — there are no `mtp.` tensors, and all 30 `conv1d.weight` tensors already have `shape[-1] == 1` — so the shift should be a no-op. `mlx-vlm` evaluates the same predicate, gets `False`, and leaves the norms alone → coherent output. `mlx-lm` shifts anyway → the RMSNorm gains are all off by +1.0 → garbage from token 1.

**Confirmation / fix.** Subtracting `1.0` back from exactly those 100 norm tensors after `mlx_lm.load()` makes generation coherent:

```
What is the capital of Japan?  ->  "The capital of Japan is **Tokyo**."
What is 17 times 23?           ->  "To find the product of 17 and 23, we can use the standard multiplication method..."
```

So the fix is to make `should_shift_norm_weights` agree with the actual checkpoint state (the un-shifted, non-MTP, already-sanitized-conv1d case must not shift). This also explains why the sibling `Qwen3.5-35B-A3B` was fine — its norm weights happen not to trip the predicate. Happy to open a PR.

**Comment by PhilipJohnBasile** [comment]
A structural fix with regression coverage is now in draft #1623. It removes MTP presence from the format predicate and applies the RMSNorm `+1` conversion only when raw Conv1D layout proves the checkpoint is unsanitized, matching the earlier maintainer feedback on #1198. The focused dense+MoE regression passes. A local raw Qwen3.6 base checkpoint also validates the intended positive case: norm mean `-0.0238` with Conv1D shape `(10240, 1, 4)`, so raw conversion still shifts while already-converted `(…, …, 1)` checkpoints do not.

**Comment by avlp12** [comment]
Confirming this on a dense model, in case it helps close out the diagnosis: the reports here are `qwen3_5_moe` 35B checkpoints, and I hit the identical failure on `Qwen/Qwen3.8-27B` (dense `qwen3_5`, 64 layers, hybrid GatedDeltaNet/full-attention) quantized to 4-bit.

Same signature as @raullenchai found — the loader, not the forward pass. Teacher-forced NLL on a fixed Korean passage, same weights and prompt:

| mlx-lm | NLL |
| --- | --- |
| current `main` | 17.460 |
| with the predicate fix | 1.679 |

17.46 is worse than uniform over a 248k vocab, so the logits are actively wrong. `mlx_lm.generate` produces fluent text once patched.

The trigger is narrower than "VLM checkpoint", which may help anyone else landing here: **it is the MTP head, not the vision tower.** My builds without an MTP head loaded and measured correctly the whole time; only the MTP-preserving ones broke. Most Qwen 3.5/3.6 VLM checkpoints happen to ship both, which is why it presents as a VLM problem. A text-only checkpoint that keeps its MTP tensors is affected too.

That also makes it a bad failure to own: it looks like the quantization is broken, not the loader, so the natural reaction is to rebuild. I re-quantized twice before finding this.

Workarounds until a fix lands: load through `mlx-vlm`, or strip the `mtp.` tensors from the checkpoint, or apply the one-line predicate change. The fix is in flight in #1623 (and #1735, which adds a regression case for a converted checkpoint that still carries MTP tensors — the existing test strips them before re-sanitizing, so it cannot catch this).
